feat: Add GroupColumn `Decimal128Array`

jonathanc-n opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #13505.

## Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

## What changes are included in this PR?
Added group column for `Decimal128Array`
<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

## Are these changes tested?
Yes, slt tests.
<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

## Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

jayzhan211 commented on datafusion/physical-plan/src/aggregates/group_values/multi_group_by/primitive.rs:
```diff
@@ -190,9 +191,13 @@ impl<T: ArrowPrimitiveType, const NULLABLE: bool> GroupColumn
+
         // Set timezone information for timestamp
-        Arc::new(arr.with_data_type(data_type))
+        adjust_output_array(&data_type, array_ref)
```
If we need `adjust_output_array` then I think we don't need `with_data_type` to set the timezone information since it is handled  by `adjust_output_array` as well

jonathanc-n replied:
The precision and scale aren't kept in the generic when constructing the buffer, so i think i need t keep with_data_type

jayzhan211 replied:
It is kept in `data_type`. When you run `adjust_output_array`, the precision of Decimal should be set with `with_precision_and_scale`.

```rust
pub fn adjust_output_array(data_type: &DataType, array: ArrayRef) -> Result<ArrayRef> {
    let array = match data_type {
        DataType::Decimal128(p, s) => Arc::new(
            array
                .as_primitive::<Decimal128Type>()
                .clone()
                .with_precision_and_scale(*p, *s)?,
        ) as ArrayRef,
```

jonathanc-n replied:
Yeah that is what I thought as well but in https://github.com/apache/datafusion/pull/13564/commits/ea6f77aff4ae52faef6adf961697bb589820f844 it fails due to precision/scale errors due to the original array not having `.with_data_type` on it

jayzhan211 replied:
The error you mentioned can be fixed by adding support for Decimal256

alamb replied:
I tried backing out the use of adjust_data_type and I didn't see any failures locally 🤔

alamb reviewed (commented):
Thank you for working on this @jonathanc-n  -- your code seems good but something is going wonky with the data types:
1. the fuzz test is failing intermittently
2. The extra adjustment of output type seems unecessary 🤔

alamb commented:
Here is a PR to deprecate `adjust_output_array`: https://github.com/apache/datafusion/pull/13585

jonathanc-n commented:
@alamb thanks for the tip, I reverted that change! For the `test_sum` fuzz test, I removed it the same way I did for the Float types due to casting to a DateType. This was the error I got after executing with a backtrace (many more of this same error):
`ERROR: Cast error: Failed to convert 39087111289254881.41 to datetime for Timestamp(Millisecond, None)`

jayzhan211 commented:
> For the test_sum fuzz test, I removed it the same way I did for the Float types due to casting to a DateType

I didn't see the removal you mentioned, but the change looks better to me now

alamb commented on datafusion/core/tests/fuzz_cases/aggregation_fuzzer/data_generator.rs:
```diff
@@ -87,7 +87,12 @@ impl DatasetGeneratorConfig {
             .filter_map(|d| {
                 if d.column_type.is_numeric()
-                    && !matches!(d.column_type, DataType::Float32 | DataType::Float64)
+                    && !matches!(
```
From my perspective other than this change, this PR is ready to go.

Thank you @jonathanc-n

jonathanc-n replied:
> @alamb thanks for the tip, I reverted that change! For the `test_sum` fuzz test, I removed it the same way I did for the Float types due to casting to a DateType. This was the error I got after executing with a backtrace (many more of this same error): `ERROR: Cast error: Failed to convert 39087111289254881.41 to datetime for Timestamp(Millisecond, None)`

I'm getting this, should I still remove it?

jayzhan211 replied:
The real issue is that we have decimal(38, 10) which is the fixed precision for `sum` and it mismatches with the fuzz test which has random precision

jayzhan211 replied:
```
statement ok
create table t as values (arrow_cast(21, 'Decimal128(20, 3)'), 1), (arrow_cast(11, 'Decimal128(20, 3)'), 2);

query RR
select sum(column1), sum(distinct column1) from t group by column2;
----
11 11
21 21

query TT
select arrow_typeof(sum(column1)), arrow_typeof(sum(distinct column1)) from t group by column2;
----
Decimal128(38, 3) Decimal128(30, 3)
Decimal128(38, 3) Decimal128(30, 3)
```

jayzhan211 replied:
Decimal128(38, 3) for normal precision, Decimal128(30, 3) for grouping. Not sure why there is mismatch in fuzz test. We should either align the precision for both cases or fix the fuzz schema check if they are not necessary to have the same precision like slt

```
query TT
select arrow_typeof(sum(column1)), arrow_typeof(sum(distinct column1)) from t group by column2;
----
Decimal128(38, 3) Decimal128(30, 3)
Decimal128(38, 3) Decimal128(30, 3)

query TT
explain select sum(column1), sum(distinct column1) from t group by column2;
----
logical_plan
01)Projection: sum(alias2) AS sum(t.column1), sum(alias1) AS sum(DISTINCT t.column1)
02)--Aggregate: groupBy=[[t.column2]], aggr=[[sum(alias2), sum(alias1)]]
03)----Aggregate: groupBy=[[t.column2, t.column1 AS alias1]], aggr=[[sum(t.column1) AS alias2]]
04)------TableScan: t projection=[column1, column2]
physical_plan
01)ProjectionExec: expr=[sum(alias2)@1 as sum(t.column1), sum(alias1)@2 as sum(DISTINCT t.column1)]
02)--AggregateExec: mode=FinalPartitioned, gby=[column2@0 as column2], aggr=[sum(alias2), sum(alias1)]
03)----CoalesceBatchesExec: target_batch_size=8192
04)------RepartitionExec: partitioning=Hash([column2@0], 4), input_partitions=4
05)--------AggregateExec: mode=Partial, gby=[column2@0 as column2], aggr=[sum(alias2), sum(alias1)]
06)----------AggregateExec: mode=FinalPartitioned, gby=[column2@0 as column2, alias1@1 as alias1], aggr=[alias2]
07)------------CoalesceBatchesExec: target_batch_size=8192
08)--------------RepartitionExec: partitioning=Hash([column2@0, alias1@1], 4), input_partitions=4
09)----------------RepartitionExec: partitioning=RoundRobinBatch(4), input_partitions=1
10)------------------AggregateExec: mode=Partial, gby=[column2@1 as column2, column1@0 as alias1], aggr=[alias2]
11)--------------------MemoryExec: partitions=1, partition_sizes=[1]
```

alamb commented:
> @alamb thanks for the tip, I reverted that change! For the `test_sum` fuzz test, I removed it the same way I did for the Float types due to casting to a DateType. This was the error I got after executing with a backtrace (many more of this same error): `ERROR: Cast error: Failed to convert 39087111289254881.41 to datetime for Timestamp(Millisecond, None)`

I think I found the issue (thanks @jayzhan211 for the sluthing):
- https://github.com/jonathanc-n/datafusion/pull/1

(also shout out to @Rachelint  for writing the first version of these fuzz testers. We just avoided a potential bug with it!)

jonathanc-n commented:
@alamb @jayzhan211 Thanks for the reviews and changes!

jayzhan211 reviewed (approved):
👍🏻

alamb reviewed (approved):
Thanks @jonathanc-n
